taichi/backends/vulkan/vulkan_device.cpp

```diff
@@ -1514,14 +1525,18 @@ void VulkanStream::submit(CommandList *cmdlist_) {
   submit_info.commandBufferCount = 1;
   submit_info.pCommandBuffers = &buffer->buffer;
 
-  if (last_semaphore_) {
-    submit_info.waitSemaphoreCount = 1;
-    submit_info.pWaitSemaphores = &last_semaphore_->semaphore;
-    submit_info.pWaitDstStageMask = &stage_flag;
+  std::vector<VkSemaphore> vk_wait_semaphores;
+
+  for (const StreamSemaphore &sema_ : wait_semaphores) {
+    auto sema = std::static_pointer_cast<VulkanStreamSemaphoreObject>(sema_);
+    vk_wait_semaphores.push_back(sema->semaphore->semaphore);
```

Comment:
`sema->semaphore->semaphore` is a bit confusing :P

Reply:
Yeah, the inner one could probably be something like `rep`, `ptr`, `raw`, ..

Reply:
The innermost one is under `vkapi` i.e. our shared ptr Vulkan wrapper, most RHIs have the tradition of using object names for those. The center one should probably be something like `vk_ref`...